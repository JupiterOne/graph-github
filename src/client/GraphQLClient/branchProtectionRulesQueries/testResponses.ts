const branchProtectionRulesResponses = [
  {
    repository: {
      repoId: 'R_kgDOI2pL-A',
      name: 'reimagined-barnacle',
      branchProtectionRules: {
        nodes: [
          {
            id: 'BPR_sdflk_slkwefoi',
            requiresLinearHistory: false,
            requiredApprovingReviewCount: 2,
            dismissesStaleReviews: false,
            requiresCodeOwnerReviews: false,
            requiresCommitSignatures: true,
            isAdminEnforced: false,
            allowsForcePushes: false,
            allowsDeletions: false,
            blocksCreations: false,
            requiresConversationResolution: false,
            pattern: 'release-*',
            requiresApprovingReviews: true,
            requiredStatusCheckContexts: [],
            creator: {
              login: 'VDubber',
            },
            databaseId: 25221868,
            requiresStatusChecks: false,
            requiresStrictStatusChecks: true,
            restrictsPushes: false,
            restrictsReviewDismissals: false,
            requiredStatusChecks: [],
            bypassPullRequestAllowances: {
              nodes: [
                {
                  actor: {
                    __typename: 'Team',
                    id: 'T_kwDOBfl5Zc4AWFta',
                    name: 'j1-nested-team',
                  },
                },
                {
                  actor: {
                    __typename: 'User',
                    id: 'MDQ6VXNlcjU1NzcxODIz',
                    login: 'adam-in-ict',
                    email: '',
                  },
                },
              ],
            },
          },
          {
            requiresLinearHistory: false,
            requiredApprovingReviewCount: 1,
            dismissesStaleReviews: true,
            requiresCodeOwnerReviews: false,
            requiresCommitSignatures: false,
            isAdminEnforced: true,
            allowsForcePushes: false,
            allowsDeletions: false,
            blocksCreations: false,
            requiresConversationResolution: false,
            pattern: 'main',
            requiresApprovingReviews: true,
            requiredStatusCheckContexts: [],
            creator: {
              login: 'electricgull',
            },
            databaseId: 27924631,
            requiresStatusChecks: true,
            requiresStrictStatusChecks: false,
            restrictsPushes: false,
            restrictsReviewDismissals: false,
            requiredStatusChecks: [],
            bypassPullRequestAllowances: {
              nodes: [
                {
                  actor: {
                    __typename: 'Team',
                    id: 'T_kwDOBfl5Zc4AWFta',
                    name: 'j1-nested-team',
                  },
                },
                {
                  actor: {
                    __typename: 'User',
                    id: 'MDQ6VXNlcjE2NTYyMDg1',
                    login: 'electricgull',
                    email: '',
                  },
                },
              ],
            },
          },
        ],
      },
    },
  },
];

export { branchProtectionRulesResponses };
