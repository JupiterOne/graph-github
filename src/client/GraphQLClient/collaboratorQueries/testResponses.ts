const repoCollaborators = [
  {
    repository: {
      id: 'R_kgDOG_MJdQ',
      collaborators: {
        edges: [
          {
            node: {
              id: 'MDQ6VXNlcjU5MjMwMDI=',
              name: 'Samuel Poulton',
              login: 'VDubber',
            },
            permission: 'ADMIN',
          },
          {
            node: {
              id: 'U_kgDOBgq4WA',
              name: null,
              login: 'spoultonV2',
            },
            permission: 'READ',
          },
        ],
        pageInfo: {
          endCursor: 'more2come==',
          hasNextPage: true,
        },
      },
    },
    rateLimit: {
      cost: 2,
    },
  },
  {
    repository: {
      id: 'R_kgDOG_MJdQ',
      collaborators: {
        edges: [
          {
            node: {
              id: 'MDQ6VXNlcjU5MjMwMDI=',
              name: 'Harper Smith',
              login: 'HSM',
            },
            permission: 'ADMIN',
          },
          {
            node: {
              id: 'U_kgDOBgq4WA',
              name: 'Unkonw 2',
              login: 'u2',
            },
            permission: 'READ',
          },
        ],
        pageInfo: {
          endCursor: 'more2come==',
          hasNextPage: false,
        },
      },
    },
    rateLimit: {
      cost: 4,
    },
  },
];

export { repoCollaborators };
