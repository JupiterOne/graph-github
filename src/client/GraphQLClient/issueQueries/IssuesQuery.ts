import {
  BaseQueryState,
  BuildQuery,
  CursorState,
  IssueResponse,
  IteratePagination,
  ProcessResponse,
} from '../types';
import { ExecutableQuery } from '../CreateQueryExecutor';
import paginate, { MAX_REQUESTS_LIMIT } from '../paginate';
import utils from '../utils';
import fragments from '../fragments';

interface QueryState extends BaseQueryState {
  issues: CursorState;
}

type QueryParams = {
  repoName: string;
  login: string;
  lastExecutionTime: string;
  maxLimit: number;
};

const MAX_RESOURCES_PER_EXECUTION = 500;

/**
 * Builds query for searching for applicable Issues.
 * Pagination is handled at the root level but not
 * currently for inner resources.
 *
 * @param queryParams
 * @param queryState
 */
const buildQuery: BuildQuery<QueryParams, QueryState> = (
  queryParams,
  queryState,
): ExecutableQuery => {
  const query = `
    query(
      $repoName: String!,
      $login: String!
      $since: DateTime!,
      $maxSearchLimit: Int!, 
      $maxInnerLimit: Int!,
      $issuesCursor: String
    ) {
      repository(name: $repoName, owner: $login) {
        id
        name
        issues(first: $maxSearchLimit, after: $issuesCursor, filterBy: { since: $since }) {
          nodes {
            ${fragments.issueFields}
            assignees(first: $maxInnerLimit) {
              nodes {
                name
                login
              }
            }
            labels(first: $maxInnerLimit) {
              nodes {
                id
                name
              }
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
      }
      ...${fragments.rateLimit}
    }`;

  return {
    query,
    ...(queryState?.rateLimit && {
      rateLimit: queryState.rateLimit,
    }),
    queryVariables: {
      login: queryParams.login,
      repoName: queryParams.repoName,
      since: queryParams.lastExecutionTime,
      maxSearchLimit: queryParams.maxLimit,
      maxInnerLimit: MAX_REQUESTS_LIMIT,
      ...(queryState?.issues?.hasNextPage && {
        issuesCursor: queryState?.issues.endCursor,
      }),
    },
  };
};

/**
 * Processes response data into a resource object
 * ready to be passed to the iterator.
 * @param responseData
 * @param iteratee
 */
const processResponseData: ProcessResponse<IssueResponse, QueryState> = async (
  responseData,
  iteratee,
) => {
  if (!responseData) {
    throw new Error('responseData param is required');
  }

  console.log('Executed query for repo collaborators');

  const rateLimit = responseData.rateLimit;
  const issues = responseData.repository?.issues?.nodes ?? [];

  for (const issue of issues) {
    if (!utils.hasProperties(issue)) {
      continue;
    }

    const resource = {
      ...issue,
      repoId: responseData.repository?.id,
      repoName: responseData.repository?.name,
      assignees: issue.assignees?.nodes ?? [],
      labels: issue.labels?.nodes ?? [],
    };

    await iteratee(resource);
  }

  return {
    rateLimit,
    issues: responseData.repository?.issues?.pageInfo,
  };
};

/**
 * Iterate issues, with pagination, up to 500.
 * Utilizes queryParams.lastExecutionTime to query the minimal amount.
 * @param queryParams
 * @param iteratee
 * @param execute
 */
const iterateIssues: IteratePagination<QueryParams, IssueResponse> = async (
  queryParams,
  execute,
  iteratee,
  logger,
) => {
  return paginate(
    queryParams,
    iteratee,
    execute,
    buildQuery,
    processResponseData,
    (queryState, issuesFetched) =>
      (!queryState?.issues?.hasNextPage ?? true) ||
      issuesFetched >= MAX_RESOURCES_PER_EXECUTION,
    logger,
    'maxLimit',
  );
};

export default { iterateIssues };
