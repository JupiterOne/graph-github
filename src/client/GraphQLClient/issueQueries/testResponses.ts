const issuesResponses = [
  {
    search: {
      issueCount: 3,
      edges: [
        {
          node: {
            id: '111',
            title: 'Need to add code part 1',
            assignees: {
              totalCount: 1,
              nodes: [
                {
                  name: null,
                  login: 'spoultonV2',
                },
              ],
              pageInfo: {
                endCursor: 'Y3Vyc29yOnYyOpHOBgq4WA==',
                hasNextPage: false,
              },
            },
            labels: {
              totalCount: 9,
              nodes: [
                {
                  id: 'LA_kwDOG5ypUc7mjrvd',
                  name: 'bug',
                },
                {
                  id: 'LA_kwDOG5ypUc7mjrvi',
                  name: 'feature',
                },
              ],
              pageInfo: {
                endCursor:
                  'Y3Vyc29yOnYyOpK5MjAyMi0wMi0yNFQxMTowNDo1OC0wNzowMM7mjrwE',
                hasNextPage: false,
              },
            },
          },
        },
        {
          node: {
            id: '2222',
            title: 'Need to add code part 2',
            assignees: {
              totalCount: 1,
              nodes: [
                {
                  name: null,
                  login: 'maxmin',
                },
              ],
              pageInfo: {
                endCursor: 'Y3Vyc29yOnYyOpHOBgq4WA==',
                hasNextPage: false,
              },
            },
            labels: {
              totalCount: 9,
              nodes: [
                {
                  id: 'LA_kwDOG5ypUc7mjrvd',
                  name: 'bug',
                },
                {
                  id: 'LA_kwDOG5ypUc7mjrvi',
                  name: 'critical',
                },
              ],
              pageInfo: {
                endCursor:
                  'Y3Vyc29yOnYyOpK5MjAyMi0wMi0yNFQxMTowNDo1OC0wNzowMM7mjrwE',
                hasNextPage: false,
              },
            },
          },
        },
      ],
      pageInfo: {
        endCursor: 'Y3Vyc29yOjE=',
        hasNextPage: true,
      },
    },
  },
  // Send page
  {
    search: {
      issueCount: 3,
      edges: [
        {
          node: {
            id: '333',
            title: 'Need to add code part 3',
            assignees: {
              totalCount: 1,
              nodes: [
                {
                  name: null,
                  login: 'chris3',
                },
              ],
              pageInfo: {
                endCursor: 'Y3Vyc29yOnYyOpHOBgq4WA==',
                hasNextPage: false,
              },
            },
            labels: {
              totalCount: 9,
              nodes: [
                {
                  id: 'LA_kwDOG5ypUc7mjrvd',
                  name: 'defect',
                },
                {
                  id: 'LA_kwDOG5ypUc7mjrvi',
                  name: 'easy',
                },
              ],
              pageInfo: {
                endCursor:
                  'Y3Vyc29yOnYyOpK5MjAyMi0wMi0yNFQxMTowNDo1OC0wNzowMM7mjrwE',
                hasNextPage: false,
              },
            },
          },
        },
      ],
      pageInfo: {
        endCursor: 'Y3Vyc29yOjE=',
        hasNextPage: false,
      },
    },
    rateLimit: {
      cost: 3,
      remaining: 4983,
    },
  },
];

export { issuesResponses };
