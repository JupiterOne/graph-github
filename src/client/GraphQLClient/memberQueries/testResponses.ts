const orgMembers = [
  {
    organization: {
      id: 'O_kgDOBfl5ZQ',
      membersWithRole: {
        totalCount: 4,
        edges: [
          {
            node: {
              id: 'm_1',
              login: 'VDubber',
              name: 'Samuel Poulton',
              isSiteAdmin: false,
              company: 'JupiterOne',
              createdAt: '2013-11-12T20:20:50Z',
              databaseId: 5923002,
              email: 'samuel.poulton@jupiterone.com',
              isEmployee: false,
              location: null,
              updatedAt: '2022-03-15T15:43:54Z',
              url: 'https://github.com/VDubber',
              websiteUrl: null,
            },
            hasTwoFactorEnabled: true,
            role: 'ADMIN',
          },
          {
            node: {
              id: 'm_2',
              login: 'codestuffs',
              name: 'Jump Jack',
              isSiteAdmin: false,
              company: null,
              createdAt: '2022-03-10T21:30:13Z',
              databaseId: 101365848,
              email: 'jump@jack.com',
              isEmployee: false,
              location: null,
              updatedAt: '2022-03-10T21:30:13Z',
              url: 'https://github.com/codestuffs',
              websiteUrl: null,
            },
            hasTwoFactorEnabled: false,
            role: 'MEMBER',
          },
        ],
        pageInfo: {
          endCursor: 'more2come',
          hasNextPage: true,
        },
      },
    },
    rateLimit: {
      cost: 5,
    },
  },
  {
    organization: {
      id: 'O_kgDOBfl5ZQ',
      membersWithRole: {
        totalCount: 4,
        edges: [
          {
            node: {
              id: 'm_3',
              login: 'JP',
              name: 'Joe Poulton',
              isSiteAdmin: false,
              company: 'JupiterOne',
              createdAt: '2013-11-12T20:20:50Z',
              databaseId: 5923002,
              email: 'j.poulton@jupiterone.com',
              isEmployee: false,
              location: null,
              updatedAt: '2022-03-15T15:43:54Z',
              url: 'https://github.com/JP',
              websiteUrl: null,
            },
            hasTwoFactorEnabled: true,
            role: 'ADMIN',
          },
          {
            node: {
              id: 'm_4',
              login: 'spv2',
              name: null,
              isSiteAdmin: false,
              company: null,
              createdAt: '2022-03-10T21:30:13Z',
              databaseId: 101365848,
              email: 'spv2@j1.com',
              isEmployee: false,
              location: null,
              updatedAt: '2022-03-10T21:30:13Z',
              url: 'https://github.com/spv2',
              websiteUrl: null,
            },
            hasTwoFactorEnabled: false,
            role: 'MEMBER',
          },
        ],
        pageInfo: {
          endCursor: 'end==',
          hasNextPage: false,
        },
      },
    },
    rateLimit: {
      cost: 5,
    },
  },
];

const teamMembers = [
  {
    organization: {
      id: 'O_kgDOBfl5ZQ',
      team: {
        id: 'T_kwDOBfl5Zc4AWFtb',
        name: 'j1-ingest-eng',
        members: {
          edges: [
            {
              node: {
                id: 'm1',
                name: 'Joe Poulton',
                login: 'JP',
              },
              role: 'MAINTAINER',
            },
            {
              node: {
                id: 'm2',
                name: 'Samuel Poulton',
                login: 'VDubber',
              },
              role: 'ADMIN',
            },
            {
              node: {
                id: 'm3',
                name: 'E',
                login: 'EP',
              },
              role: 'MAINTAINER',
            },
          ],
          pageInfo: {
            endCursor: 'more2come',
            hasNextPage: true,
          },
        },
      },
    },
    rateLimit: {
      cost: 2,
    },
  },
  {
    organization: {
      id: 'O_kgDOBfl5ZQ',
      team: {
        id: 'T_kwDOBfl5Zc4AWFtb',
        name: 'j1-ingest-eng',
        members: {
          edges: [
            {
              node: {
                id: 'm4',
                name: 'Silvia Smith',
                login: 'SSM',
              },
              role: 'MAINTAINER',
            },
          ],
          pageInfo: {
            endCursor: 'end',
            hasNextPage: false,
          },
        },
      },
    },
    rateLimit: {
      cost: 1,
    },
  },
];

export { orgMembers, teamMembers };
