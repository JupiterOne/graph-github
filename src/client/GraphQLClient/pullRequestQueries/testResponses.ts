/**
 * Multiple PR responses.
 */

const pullRequestsPublic = [
  {
    search: {
      issueCount: 3,
      edges: [
        {
          node: {
            id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
            title: 'Update CHANGELOG.md',
            closed: false,
          },
        },
        {
          node: {
            id: 'MDExOlB1bGxSZXF1ZXN0OQ==',
            title: 'Create CHANGELOG.md #2',
            closed: false,
            mergeCommit: {
              id: 'C_kwDOIAVVA9oAKDg1NjNlOWIyZDYzMzRiNzUwZjhmN2I2MzA4NTgxODg5NTBkMmFiOTI',
              oid: '8563e9b2d6334b750f8f7b630858188950d2ab92',
              associatedPullRequests: {
                nodes: [
                  {
                    number: 2,
                    id: 'PR_kwDOIAVVA84_DshD',
                    state: 'MERGED',
                    reviewDecision: null,
                    url: 'https://github.com/j1-ingest/ubiquitous-umbrella/pull/2',
                  },
                ],
              },
            },
          },
        },
      ],
      pageInfo: {
        endCursor: 'Y3Vyc29yOjI=',
        hasNextPage: true,
      },
    },
    rateLimit: {
      cost: 2,
    },
  },
  {
    search: {
      issueCount: 3,
      edges: [
        {
          node: {
            id: 'pr3',
            title: 'Update CHANGELOG.md - page 2',
            closed: false,
          },
        },
      ],
      pageInfo: {
        endCursor: 'zxcvee=',
        hasNextPage: false,
      },
    },
    rateLimit: {
      cost: 1,
    },
  },
];

const pullRequestsPublicInnerPagination = {
  search: {
    issueCount: 3,
    edges: [
      {
        node: {
          id: 'MDExOlB1bGxSZXF1ZXN0OQ==',
          title: 'pullRequestsPublicInnerPagination',
          headRepository: {
            name: 'musical-palm-tree',
            owner: {
              login: 'J1-Test',
            },
          },
        },
      },
    ],
    pageInfo: {
      endCursor: 'Y3Vyc29yOjI=',
      hasNextPage: false,
    },
  },
  rateLimit: {
    cost: 1,
  },
};

const emptyPullRequest = [
  {
    search: {
      issueCount: 0,
      edges: [],
      pageInfo: {
        endCursor: null,
        hasNextPage: false,
      },
    },
    rateLimit: {
      cost: 1,
    },
  },
];

export {
  pullRequestsPublic,
  pullRequestsPublicInnerPagination,
  emptyPullRequest,
};

/**
 * Single Query Responses
 */

const singleQueryFullResponse = {
  repository: {
    pullRequest: {
      id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
      title: 'Update CHANGELOG.md',
      mergeCommit: {
        id: 'C_kwDOIAVVA9oAKDg1NjNlOWIyZDYzMzRiNzUwZjhmN2I2MzA4NTgxODg5NTBkMmFiOTI',
        associatedPullRequests: {
          nodes: [
            {
              number: 2,
              id: 'PR_kwDOIAVVA84_DshD',
              state: 'MERGED',
              reviewDecision: null,
              url: 'https://github.com/j1-ingest/ubiquitous-umbrella/pull/2',
            },
          ],
        },
      },
      commits: {
        totalCount: 1,
        nodes: [
          {
            commit: {
              id: 'commit1',
            },
          },
        ],
        pageInfo: {
          endCursor: 'MQ',
          hasNextPage: false,
        },
      },
      reviews: {
        totalCount: 2,
        nodes: [
          {
            id: 'review1',
          },
          {
            id: 'review2',
          },
        ],
        pageInfo: {
          endCursor: 'Y3Vyc29yOnYyOpO0M',
          hasNextPage: true,
        },
      },
      labels: {
        totalCount: 4,
        nodes: [
          {
            id: 'label1=',
            name: 'bug',
          },
          {
            id: 'label2=',
            name: 'documentation',
          },
        ],
        pageInfo: {
          endCursor: 'Y3Vyc2==',
          hasNextPage: true,
        },
      },
    },
  },
  rateLimit: {
    limit: 5000,
    cost: 1,
    remaining: 4998,
    resetAt: '2022-03-07T22:04:07Z',
  },
};

const singleQueryWithPartialInnerResources = {
  repository: {
    pullRequest: {
      id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
      title: 'singleQueryWithPartialInnerResources',
      commits: {
        totalCount: 1,
        nodes: [
          {
            commit: {
              id: 'MDY6',
            },
          },
        ],
        pageInfo: {
          endCursor: 'MQ',
          hasNextPage: false,
        },
      },
      labels: {
        totalCount: 3,
        nodes: [
          {
            id: 'MDU6TGFiZWwxODE=',
            name: 'bug',
          },
          {
            id: 'MDU6TGFiZWwxODI=',
            name: 'documentation',
          },
        ],
        pageInfo: {
          endCursor: 'Y3Vyc2==',
          hasNextPage: true,
        },
      },
    },
  },
  rateLimit: {
    limit: 5000,
    cost: 1,
    remaining: 4998,
    resetAt: '2022-03-07T22:04:07Z',
  },
};

const singleQueryInnerResourcePaginationComplete = {
  repository: {
    pullRequest: {
      id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
      title: 'innerResourcePaginationComplete',
      commits: {
        totalCount: 1,
        nodes: [],
        pageInfo: {
          endCursor: null,
          hasNextPage: false,
        },
      },
      reviews: {
        totalCount: 3,
        nodes: [
          {
            id: 'review3==',
          },
        ],
        pageInfo: {
          endCursor: 'bowek',
          hasNextPage: false,
        },
      },
      labels: {
        totalCount: 4,
        nodes: [
          {
            id: 'label3=',
            name: 'first-task',
          },
          {
            id: 'label4=',
            name: 'feature',
          },
        ],
        pageInfo: {
          endCursor: 'sdvksdf2',
          hasNextPage: false,
        },
      },
    },
  },
  rateLimit: {
    limit: 5000,
    cost: 1,
    remaining: 4999,
    resetAt: '2022-03-08T20:46:50Z',
  },
};

const commitsQueryResponse = [
  {
    repository: {
      pullRequest: {
        id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
        commits: {
          totalCount: 3,
          nodes: [
            {
              commit: {
                oid: 'f5294eb2f1545eb113d3d3d5a7f3977934a13798',
                message: 'README Update',
                authoredDate: '2023-05-15T14:05:06Z',
                author: { user: { login: 'nealajpatel' } },
              },
            },
            {
              commit: {
                oid: '3b00cc1814bc8e93d2676f48494118a339d41839',
                message: 'README Update',
                authoredDate: '2023-05-15T14:04:53Z',
                author: { user: { login: 'nealajpatel' } },
              },
            },
          ],
          pageInfo: { endCursor: 'MQ', hasNextPage: true },
        },
      },
    },
    rateLimit: {
      limit: 5000,
      cost: 2,
      remaining: 5000,
      resetAt: '2023-07-17T21:41:33Z',
    },
  },
  {
    repository: {
      pullRequest: {
        id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
        commits: {
          totalCount: 3,
          nodes: [
            {
              commit: {
                oid: 'c3d5ca5b1a7e39d8f541e815c437d04579677b67',
                message: 'test PR',
                authoredDate: '2023-07-10T20:54:45Z',
                author: { user: null },
              },
            },
          ],
          pageInfo: { endCursor: null, hasNextPage: false },
        },
      },
    },
    rateLimit: {
      limit: 5000,
      cost: 1,
      remaining: 4998,
      resetAt: '2023-07-17T21:41:33Z',
    },
  },
];

const labelsQueryResponse = [
  {
    repository: {
      pullRequest: {
        id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
        labels: {
          totalCount: 1,
          nodes: [{ name: 'dependencies' }, { name: 'automation' }],
          pageInfo: { endCursor: 'MQ', hasNextPage: true },
        },
      },
    },
    rateLimit: {
      limit: 5000,
      cost: 2,
      remaining: 5000,
      resetAt: '2023-07-17T22:41:34Z',
    },
  },
  {
    repository: {
      pullRequest: {
        id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
        labels: {
          totalCount: 1,
          nodes: [{ name: 'release' }],
          pageInfo: { endCursor: null, hasNextPage: false },
        },
      },
    },
    rateLimit: {
      limit: 5000,
      cost: 1,
      remaining: 4998,
      resetAt: '2023-07-17T22:41:34Z',
    },
  },
];

const reviewsQueryResponse = [
  {
    repository: {
      pullRequest: {
        id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
        reviews: {
          totalCount: 1,
          nodes: [
            {
              commit: { oid: 'a4c090f614f3b3ed4707c549bcb2d4d5afa76805' },
              author: { name: 'Neal Patel', login: 'nealajpatel' },
              state: 'APPROVED',
            },
            {
              commit: null,
              author: { name: 'Neal Patel', login: 'nealajpatel' },
              state: 'APPROVED',
            },
          ],
          pageInfo: {
            endCursor:
              'Y3Vyc29yOnYyOpO0MjAyMy0wNS0yMFQwMzo0OTo1Nlq0MjAyMy0wNS0yMFQwMzo0OTo1NlrOVY2pjA==',
            hasNextPage: true,
          },
        },
      },
    },
    rateLimit: {
      limit: 5000,
      cost: 2,
      remaining: 5000,
      resetAt: '2023-07-17T22:41:34Z',
    },
  },
  {
    repository: {
      pullRequest: {
        id: 'MDExOlB1bGxSZXF1ZXN0MTA=',
        reviews: {
          totalCount: 1,
          nodes: [
            {
              commit: null,
              author: { name: 'Neal Patel', login: 'nealajpatel' },
              state: 'APPROVED',
            },
          ],
          pageInfo: {
            endCursor: null,
            hasNextPage: false,
          },
        },
      },
    },
    rateLimit: {
      limit: 5000,
      cost: 1,
      remaining: 4998,
      resetAt: '2023-07-17T22:41:34Z',
    },
  },
];

export {
  singleQueryFullResponse,
  singleQueryWithPartialInnerResources,
  singleQueryInnerResourcePaginationComplete,
  commitsQueryResponse,
  labelsQueryResponse,
  reviewsQueryResponse,
};
