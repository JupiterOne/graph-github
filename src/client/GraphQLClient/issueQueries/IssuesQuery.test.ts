import IssuesQuery from './IssuesQuery';
import { issuesResponses } from './testResponses';

describe('IssuesQuery', () => {
  describe('#iterateIssues', () => {
    test('Pulling data out', async () => {
      const iteratee = jest.fn();
      const execute = jest
        .fn()
        .mockResolvedValueOnce(issuesResponses[0])
        .mockResolvedValueOnce(issuesResponses[1]);

      // Act
      const result = await IssuesQuery.iterateIssues(
        'J1-Test/happy-sunshine',
        '2011-10-05T14:48:00.000Z',
        iteratee,
        execute,
      );

      // Assert
      expect(result.rateLimitConsumed).toBe(6);
      expect(iteratee).toHaveBeenCalledTimes(3);
      expect(iteratee.mock.calls[0][0]).toMatchSnapshot();
      expect(iteratee.mock.calls[1][0]).toMatchSnapshot();
      expect(iteratee.mock.calls[2][0]).toMatchSnapshot();
      expect(execute).toHaveBeenCalledTimes(2);
      expect(execute.mock.calls[0][0]).toMatchSnapshot();
      expect(execute.mock.calls[1][0]).toMatchSnapshot();
    });
  });
});
