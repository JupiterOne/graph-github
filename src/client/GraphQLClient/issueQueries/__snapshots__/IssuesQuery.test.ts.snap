// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesQuery #iterateIssues Pulling data out 1`] = `
Object {
  "assignees": Array [
    Object {
      "login": "spoultonV2",
      "name": null,
    },
  ],
  "id": "111",
  "labels": Array [
    Object {
      "id": "LA_kwDOG5ypUc7mjrvd",
      "name": "bug",
    },
    Object {
      "id": "LA_kwDOG5ypUc7mjrvi",
      "name": "feature",
    },
  ],
  "title": "Need to add code part 1",
}
`;

exports[`IssuesQuery #iterateIssues Pulling data out 2`] = `
Object {
  "assignees": Array [
    Object {
      "login": "maxmin",
      "name": null,
    },
  ],
  "id": "2222",
  "labels": Array [
    Object {
      "id": "LA_kwDOG5ypUc7mjrvd",
      "name": "bug",
    },
    Object {
      "id": "LA_kwDOG5ypUc7mjrvi",
      "name": "critical",
    },
  ],
  "title": "Need to add code part 2",
}
`;

exports[`IssuesQuery #iterateIssues Pulling data out 3`] = `
Object {
  "assignees": Array [
    Object {
      "login": "chris3",
      "name": null,
    },
  ],
  "id": "333",
  "labels": Array [
    Object {
      "id": "LA_kwDOG5ypUc7mjrvd",
      "name": "defect",
    },
    Object {
      "id": "LA_kwDOG5ypUc7mjrvi",
      "name": "easy",
    },
  ],
  "title": "Need to add code part 3",
}
`;

exports[`IssuesQuery #iterateIssues Pulling data out 4`] = `
Object {
  "query": "
    query(
      $issueQuery: String!, 
      $maxSearchLimit: Int!, 
      $maxInnerLimit: Int!,
      $issuesCursor: String
    ) {
      search(
        first: $maxSearchLimit, 
        after: $issuesCursor, 
        type: ISSUE, 
        query: $issueQuery
        ) {
          issueCount
          edges {
            node {
            ... on Issue {
    id
    activeLockReason
    author {
      ...on User {
        name
        login
      }
    }
    authorAssociation
    body
    # bodyHTML
    # bodyResourcePath
    bodyText
    bodyUrl
    closed # boolean
    closedAt
    # comments # probably a child object if we want these
    createdAt
    createdViaEmail # boolean
    databaseId
    isPinned # boolean
    lastEditedAt
    locked # boolean
    # milestone # a Milestone object, could put the fields in-line like author
    number
    # participants # Participants objects
    # projectCards # ProjectCardConnection!
    publishedAt
    # reactionGroups # : [ReactionGroup!]
    resourcePath
    state
    title
    titleHTML
    updatedAt
    url
  }
            
            ... on Issue {
                assignees(first: $maxInnerLimit) {
                  totalCount
                  nodes {
                    name
                    login
                  }
                  pageInfo {
                    endCursor
                    hasNextPage
                  }
                }
              }
              
            ... on Issue {
                labels(first: $maxInnerLimit) {
                  totalCount
                  nodes {
                    id
                    name
                  }
                  pageInfo {
                    endCursor
                    hasNextPage
                  }
                }
              }
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...on Query {
    rateLimit {
      limit
      cost
      remaining
      resetAt
    }
  }
    }",
  "queryVariables": Object {
    "issueQuery": "is:issue repo:J1-Test/happy-sunshine updated:>=2011-10-05T14:48:00.000Z",
    "maxInnerLimit": 100,
    "maxSearchLimit": 25,
  },
}
`;

exports[`IssuesQuery #iterateIssues Pulling data out 5`] = `
Object {
  "query": "
    query(
      $issueQuery: String!, 
      $maxSearchLimit: Int!, 
      $maxInnerLimit: Int!,
      $issuesCursor: String
    ) {
      search(
        first: $maxSearchLimit, 
        after: $issuesCursor, 
        type: ISSUE, 
        query: $issueQuery
        ) {
          issueCount
          edges {
            node {
            ... on Issue {
    id
    activeLockReason
    author {
      ...on User {
        name
        login
      }
    }
    authorAssociation
    body
    # bodyHTML
    # bodyResourcePath
    bodyText
    bodyUrl
    closed # boolean
    closedAt
    # comments # probably a child object if we want these
    createdAt
    createdViaEmail # boolean
    databaseId
    isPinned # boolean
    lastEditedAt
    locked # boolean
    # milestone # a Milestone object, could put the fields in-line like author
    number
    # participants # Participants objects
    # projectCards # ProjectCardConnection!
    publishedAt
    # reactionGroups # : [ReactionGroup!]
    resourcePath
    state
    title
    titleHTML
    updatedAt
    url
  }
            
            ... on Issue {
                assignees(first: $maxInnerLimit) {
                  totalCount
                  nodes {
                    name
                    login
                  }
                  pageInfo {
                    endCursor
                    hasNextPage
                  }
                }
              }
              
            ... on Issue {
                labels(first: $maxInnerLimit) {
                  totalCount
                  nodes {
                    id
                    name
                  }
                  pageInfo {
                    endCursor
                    hasNextPage
                  }
                }
              }
            }
          }
          pageInfo {
            endCursor
            hasNextPage
          }
        }
        ...on Query {
    rateLimit {
      limit
      cost
      remaining
      resetAt
    }
  }
    }",
  "queryVariables": Object {
    "issueQuery": "is:issue repo:J1-Test/happy-sunshine updated:>=2011-10-05T14:48:00.000Z",
    "issuesCursor": "Y3Vyc29yOjE=",
    "maxInnerLimit": 100,
    "maxSearchLimit": 25,
  },
}
`;
